Get all tests passing on linux arm64

benjyw opened the pull request:
1) Add terraform versions for Linux arm64
2) Switches spectral to be an NpxTool, since it has no published binary for arm64
3) Various other test tweaks

Once we have resources to run Linux arm64 CI we can start releasing wheels for that platform.

benjyw commented:
With these changes `./pants test ::` passes on Linux Arm64 with Python 3.7 and 3.8 (3.9 has a dynamic linking issue that I will look into separately).

`./cargo test` also passes.

benjyw commented on src/python/pants/backend/openapi/lint/spectral/subsystem.py:
```diff
@@ -3,31 +3,16 @@
-        "linux_arm64": "linux",
-        "linux_x86_64": "linux",
-    }
+    default_version = "@stoplight/spectral-cli@6.5.1"
```
This is an experimental backend, so we can break the semantics of this version string, and note that this backend will only be available in 2.15, and this change will be in place in 2.16, so the incompatibility window is narrow.

benjyw commented on src/python/pants/backend/python/goals/coverage_py_integration_test.py:
```diff
@@ -315,8 +315,10 @@ def test_coverage_html_xml_json_lcov(batched: bool) -> None:
 def test_default_coverage_issues_12390() -> None:
     # N.B.: This ~replicates the repo used to reproduce this issue at
     # https://github.com/alexey-tereshenkov-oxb/monorepo-coverage-pants.
-    if platform.system() == "Darwin" and platform.processor() == "arm":
```
It turns out that `platform.processor()` is an empty string on Linux ARM64, and in general `machine()` is preferred.

jsirois commented on src/python/pants/backend/python/packaging/pyoxidizer/rules_integration_test.py:
```diff
@@ -1,13 +1,22 @@
 
+skip_on_linux_arm = pytest.mark.skipif(
+    platform.system() == "Linux" and platform.machine() == "aarch64",
+    reason="PyOxidizer is not supported on Linux ARM",
```
That seems surprising.

benjyw replied:
I was surprised too, but https://pyoxidizer.readthedocs.io/en/pyoxidizer-0.17/pyoxidizer_getting_started.html:

```
PyOxidizer is officially supported on the following operating systems:

- Windows x86 (32-bit)
- Windows x86_64/amd64 (64-bit)
- macOS x86_64 (Intel processors)
- macOS aarch64 (ARM/Apple processors)
- Linux i686 (32-bit)
- Linux x86_64 (64-bit)
```

jsirois commented on src/python/pants/backend/terraform/tool.py:
```diff
@@ -32,80 +34,136 @@ class TerraformTool(TemplatedExternalTool):
             "1.0.7|macos_x86_64|80ae021d6143c7f7cbf4571f65595d154561a2a25fd934b7a8ccc1ebf3014b9b|33020029",
             "1.0.7|macos_arm64|cbab9aca5bc4e604565697355eed185bb699733811374761b92000cc188a7725|32071346",
             "1.0.7|linux_x86_64|bc79e47649e2529049a356f9e60e06b47462bf6743534a10a4c16594f443be7b|32671441",
-            "1.0.6|macos_x86_64|3a97f2fffb75ac47a320d1595e20947afc8324571a784f1bd50bd91e26d5648c|33022053",
```
Why is dropping 1.0.6 ok? It looks like you want newer versions with all 4 platforms supported but doesn't this needlessly break existing 1.0.6 users?

benjyw replied:
Ooops, that was an oversight. Fixed.

Eric-Arellano reviewed (commented):
w00t

thejcannon commented:
What's this `pubring.kbx` at the repo root? :monocle_face:

benjyw commented:
Eeeerk, I thought I had deleted that. 🤦  That was an artifact of connecting to the temporary EC2 instance I was using for testing.  I have deleted the underlying instance and key pairs, and this is a public key box, so not a security issue. Will delete these from the repo.

benjyw commented:
https://github.com/pantsbuild/pants/pull/17707
